InnerEye/ML/Histopathology/models/deepmil.py

```diff
@@ -307,6 +322,20 @@ def test_epoch_end(self, outputs: List[Dict[str, Any]]) -> None:  # type: ignore
                 fig = plot_slide_noxy(slide, score, paths, bottom_attn, key + '_bottom', ncols=4)
                 figpath = Path(key_folder_path, f'{slide}_bottom.png')
                 fig.savefig(figpath, bbox_inches='tight')
+                if len(self.slide_dataset) > 0:
+                    slide_dict = list(filter(lambda entry: entry[SlideKey.SLIDE_ID] == slide, self.slide_dataset))[0]  # type: ignore
+                    load_image_dict(slide_dict, level=self.level, margin=0)                                            # type: ignore
+                    slide_image = slide_dict[SlideKey.IMAGE]
+                    location_bbox = slide_dict['location']
+
+                    fig = plot_slide(slide_image=slide_image, scale=1.0)
+                    figpath = Path(key_folder_path, f'{slide}_thumbnail.png')
+                    fig.savefig(figpath, bbox_inches='tight')
```

Comment:
One of the coding concepts that I find really essential is "do not repeat yourself" (DRY). It greatly improves readability and maintainability of the code, and dramatically reduce the chance of errors. Here, you are repeating the same snippet three times 332-333, 337-338, 341-342. You can define a small function that takes a figure object and a filename string.

Reply:
Thanks for your suggestion. I have added a function now.